src/main/scala/scorex/core/network/NodeViewSynchronizer.scala

```diff
@@ -226,53 +257,78 @@ class NodeViewSynchronizer[TX <: Transaction,
           s": ${spam.keys.map(encoder.encode)}")
         penalizeSpammingPeer(remote)
-        val mids = spam.keys.toSeq
-        deliveryTracker.deleteSpam(mids)
-      }
-
-      if (fm.nonEmpty) {
-        val mods = fm.values.toSeq
-        viewHolderRef ! ModifiersFromRemote(remote, typeId, mods)
       }
-  }
 
-  //local node sending object ids to remote
-  protected def requestFromLocal: Receive = {
-    case RequestFromLocal(peer, modifierTypeId, modifierIds) =>
-
-      if (modifierIds.nonEmpty) {
-        val msg = Message(requestModifierSpec, Right(modifierTypeId -> modifierIds), None)
-        peer.handlerRef ! msg
+      modifierSerializers.get(typeId) match {
+        case Some(companion) =>
+          fm.foreach { case (id, bytes) =>
+            companion.parseBytes(bytes) match {
+              case Success(mod) if !(id sameElements mod.id) =>
+                log.warn(s"Declared id ${encoder.encode(id)} is not equals to calculated one ${mod.encodedId}")
+                penalizeMisbehavingPeer(remote)
+                deliveryTracker.toUnknown(id)
+                None
+              case Failure(e) =>
+                log.warn(s"Failed to parse modifier ${encoder.encode(id)}", e)
+                penalizeMisbehavingPeer(remote)
+                deliveryTracker.toUnknown(id)
+                None
+              case Success(tx: TX@unchecked) if tx.modifierTypeId == Transaction.ModifierTypeId =>
+                viewHolderRef ! LocallyGeneratedTransaction[TX](tx)
+              case Success(pmod: PMOD@unchecked) =>
+                if (modifiersCache.contains(pmod.id) || historyReaderOpt.exists(_.contains(pmod))) {
+                  // should never be here
+                  log.error(s"Received modifier ${pmod.encodedId} that is already in cache or history.")
+                } else {
+                  historyReaderOpt match {
+                    case Some(hr) =>
+                      hr.applicableTry(pmod) match {
+                        case Failure(e) if e.isInstanceOf[MalformedModifierError] =>
+                          log.warn(s"Modifier ${pmod.encodedId} is permanently invalid", e)
+                          deliveryTracker.toInvalid(id)
+                          penalizeMisbehavingPeer(remote)
+                        case _ =>
+                          modifiersCache.put(pmod.id, pmod)
+                      }
+                    case None =>
+                      log.error("Got modifier while history reader is not ready")
+                      modifiersCache.put(pmod.id, pmod)
+
+                  }
+                }
+            }
+          }
+          if (typeId != Transaction.ModifierTypeId) {
+            modifiersCache.cleanOverfull().foreach(removed => deliveryTracker.toUnknown(removed.id))
+            viewHolderRef ! ChangedCache[PMOD, HR, ModifiersCache[PMOD, HR]](modifiersCache)
+          }
+        case None => log.error(s"Undefined serializer for modifier of type $typeId")
```

Comment:
we could move this branch to the beginning of the `switch` as it is shorter. Otherwise nobody will remember here that this `None` refers to `modifierSerializers.get(typeId)` result

Reply:
with shorted message from previous comment it is now easier to remember. Also this branch should never happen, it's better to keep it last for performance reasons